dynamic_config/dynamic_config_manager.py

```diff
@@ -0,0 +1,195 @@
+"""Component that handles dynamic configuration of the Envoy proxy."""
+
+import json
+import yaml
+from enum import Enum
+from google.protobuf import json_format
+
+from abc import ABC, abstractmethod
+import envoy.config.cluster.v3.cluster_pb2 as cluster_pb2
+import envoy.service.discovery.v3.discovery_pb2 as discovery_pb2
+import envoy.extensions.upstreams.http.v3.http_protocol_options_pb2 as http_protocol_options_pb2
+import envoy.extensions.transport_sockets.tls.v3.tls_pb2 as tls_pb2
+import nighthawk.api.configuration.cluster_config_manager_pb2 as cluster_manager_pb2
+
+import random
+
+from google.protobuf import json_format
+import argparse
+import time
+import base64
+import logging
+import os
+
+
+class DynamicConfigManager(ABC):
+  """Base class for Dynamic configuration components."""
+
+  def __init__(self):
+    """Create a DynamicConfigManager."""
+    self._exit = False
+
+  def start(self):
+    """Start periodically executing a method."""
+    while not self._exit:
+      time.sleep(self.timeBeforeNextUpdate())
+      self.execute()
+
+  def stop(self):
+    """Stop periodically executing a method."""
+    self._exit = True
+
+  @abstractmethod
+  def timeBeforeNextUpdate(self):
+    """Provide the time before the next execution."""
+    pass
+
+  @abstractmethod
+  def execute(self):
+    """Periodically execute this method."""
+    pass
+
+
+class DynamicClusterConfigManager(DynamicConfigManager):
+  """Encapsulates dynamic cluster configuration."""
+
+  class Action(Enum):
+    """Enum class for mutation actions to configuration."""
+
+    ADD = 1
+    REMOVE = 2
+
+  cluster_template_string = b"""
+    name: some_service
+    # Upstream TLS configuration.
+    transport_socket:
+      name: envoy.transport_sockets.tls
+      typed_config:
+        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
+    load_assignment:
+      cluster_name: some_service
+      # Static endpoint assignment.
+      endpoints:
+      - lb_endpoints:
+    typed_extension_protocol_options:
+      envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
+        "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
+        explicit_http_config:
+          http2_protocol_options:
+            max_concurrent_streams: 100
+  """
+  cluster_template = json_format.Parse(
+      json.dumps(yaml.load(cluster_template_string, Loader=yaml.FullLoader)), cluster_pb2.Cluster())
+
+  mutate_actions = list(Action)
+
+  def __init__(self, config: cluster_manager_pb2.DynamicClusterConfigManagerSettings):
+    """Create a DynamicClusterConfigManager."""
+    DynamicConfigManager.__init__(self)
+    self.active_config = discovery_pb2.DiscoveryResponse()
```

Comment:
Can we document all the public attributes in the class docstring (or make them private if they don't need to be public)?